File: examples/elf/add_two_numbers.ts
/**
 * Copyright 2023 University of Adelaide
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { type MeasureResult, Measuresuite } from "measuresuite";

import * as path from "path";

/**
 * our objects / shared objects we create from the ./add_two_numbers.asm
 */

const filenameElf = path.resolve("./add_two_numbers.o");
const filenameShared = path.resolve("./add_two_numbers.so");
const functionSymbol = "add_two_numbers"; // c.f. ./add_two_numbers.asm:1 'global add_two_numbers' and line 3

const arg_width = 1;
const arg_num_out = 1;
const arg_num_in = 2;

// create a Measuresuite object
const ms = new Measuresuite(arg_width, arg_num_in, arg_num_out);
ms.enableChecking();

// number of batches: each function will have at least this many time-measurements
const nob = 50;

// how many times the functionA/B are called within a loop. Short functions need bigger
// numbers in order to have sensible measurements
const batchSize = 10000;
console.debug({ filenameElf, filenameShared });

ms.loadFile({ filename: filenameShared, functionSymbol }); // symbol req'd because we internally use 'dlopen'
ms.loadFile({ filename: filenameElf }); // will use the first found symbol from the elf file

// measure
const measurementResult: MeasureResult | null = ms.measure(batchSize, nob);

if (!measurementResult) {
  // this is more or less for typescript tooling to assert, that measurementResult is not nullish
  throw new Error("measurementResult is nullish. This should not happen.");
}

if (measurementResult.functions.length !== 2) {
  throw new Error("should have two functions measured.");
}

if (measurementResult.stats.incorrect !== 0) {
  throw new Error(
    "This means, that the two functions did not compute the same result."
  );
}

// this gives the entire runtime for all batches in milliseconds... Really depends on the system
console.log(
  `The Measurement took ${measurementResult.stats.runtime}ms to complete`
);

/* lets look at the cycles: (= same as number of functions loaded)
    each of the measurements have num_batches (nob) elements
*/

const [functionARes, functionBRes] = measurementResult.cycles;
console.log(`Function A's cycles: ${functionARes.join(",")}`);
console.log(`Function B's cycles: ${functionBRes.join(",")}`);

const medianA = functionARes.sort().at(Math.floor(nob / 2)) ?? -1;
const medianB = functionBRes.sort().at(Math.floor(nob / 2)) ?? -1;

console.log(`Function A's median: ${medianA}`);
console.log(`Function B's median: ${medianB}`);

console.log(
  `Function A is${medianA < medianB ? " " : " not "}faster than Function B`
);

File: examples/elf/add_two_numbers.c
/**
 * Copyright 2023 University of Adelaide
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <assert.h>
#include <measuresuite.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

int main() {

  const char *const filename = {"./add_two_numbers.o"};
  /*
   * the c-function declaration would be something like
   * void add_two_numbers(uint64_t *out0,
   *                      const uint64_t *in0,
   *                      const uint64_t *in1) {
   *    *out0 = *in0 + *in1;
   * }
   */

  /*
   * properties of the function
   */

  // each parameter of the test function points to only one element (i.e.
  // out0[0],)
  const int arg_width = 1;

  // one out-parameter (there is only out0, no out1, etc.)
  const int arg_num_out = 1;

  // two in-paramters (in0, in1)
  const int arg_num_in = 2;

  // our measuresuite handle
  measuresuite_t ms = NULL;

  // initializing the measuresuite
  ms_initialize(&ms, arg_width, arg_num_in, arg_num_out);

  // id = -1 means 'load to the next free slot'
  // it will be set to the 'slot taken by the loaded function', after return
  int id = -1;
  ms_load_file(ms,  // handle
               ELF, // type of input data is elf
               filename,
               NULL, // symbol (ignored for BIN/ASM, optional for ELF, required
                     // for SHARED_OBJECT)
               &id); // ID (in/out)

  // first slot (subsequent loads will load to 1, 2, ...)
  assert(id == 0);

  // measure parameters
  const int number_of_batches = 10; // 10 batches of
  const int batch_size =
      100; // 100 iterations of the function-unter-test (add_two_asm), each

  // run the measurement
  ms_measure(ms, batch_size, number_of_batches);

  // prepare got JSON collection
  const char *json = NULL;
  size_t len = 0;
  ms_get_json(ms, &json, &len);

  assert(json != NULL);
  assert(len != 0);

  // prints the result json to stdout.
  printf("%s\n", json);

  // free all (internal) resources
  ms_terminate(ms);

  return 0;
}
